store/watcher_hub.go

```diff
@@ -69,13 +73,16 @@ func (wh *watcherHub) watch(key string, recursive, stream bool, index, storeInde
 		hub:        wh,
 	}
 
-	wh.mutex.Lock()
-	defer wh.mutex.Unlock()
-	// If the event exists in the known history, append the EtcdIndex and return immediately
-	if event != nil {
-		event.EtcdIndex = storeIndex
-		w.eventChan <- event
-		return w, nil
+	// If the events exists in the known history, append the EtcdIndex and return immediately unless
+	// stream=true, in which case we find all matching events, and add this watcher to the notify list.
+	if len(events) > 0 {
+		for _, event := range events {
+			event.EtcdIndex = storeIndex
+			w.eventChan <- event
```

Comment:
I want to make sure this is non-blocking since we are holding a lock when reach this line.

```go
select {
    case w.eventChan <- event:
    default:
}
```